examples/event_prop/shd_delay_shift_blend.py

```diff
@@ -0,0 +1,274 @@
+import numpy as np
+
+from ml_genn import Connection, Network, Population
+from ml_genn.callbacks import Checkpoint, SpikeRecorder, VarRecorder, Callback
+from ml_genn.compilers import EventPropCompiler, InferenceCompiler
+from ml_genn.connectivity import Dense,FixedProbability
+from ml_genn.initializers import Normal
+from ml_genn.neurons import LeakyIntegrate, LeakyIntegrateFire, SpikeInput
+from ml_genn.optimisers import Adam
+from ml_genn.serialisers import Numpy
+from ml_genn.synapses import Exponential
+from tonic import DiskCachedDataset
+from tonic.datasets import SHD
+
+from time import perf_counter
+from ml_genn.utils.data import (calc_latest_spike_time, calc_max_spikes,
+                                preprocess_tonic_spikes)
+
+from ml_genn.compilers.event_prop_compiler import default_params
+import copy
+
+NUM_HIDDEN = 256
+N_DELAY = 10
+T_DELAY = 30
+BATCH_SIZE = 32
+NUM_EPOCHS = 50
+EXAMPLE_TIME = 20.0
+AUGMENT_SHIFT = 40
+P_BLEND = 0.5
+DT = 1.0
+
+
+class EaseInSchedule(Callback):
+    def __init__(self):
+        pass
+
+    def set_params(self, compiled_network, **kwargs):
+        self._optimisers = [optimiser[0] for optimiser in compiled_network.optimisers[:3]]
+
+    def on_batch_begin(self, batch):
+        # Set parameter to return value of function
+        for optimiser in self._optimisers:
+            current_lr = getattr(optimiser, "alpha") * 1.05
+            if 0.001 > current_lr:
+                setattr(optimiser, "alpha", current_lr)
+            else:
+                setattr(optimiser, "alpha", 0.001)
+
+
+class Shift:
+    def __init__(self, f_shift, sensor_size):
+        self.f_shift = f_shift
+        self.sensor_size = sensor_size
+
+    def __call__(self, events: np.ndarray) -> np.ndarray:
+        # Shift events
+        events_copy = copy.deepcopy(events)
+        events_copy["x"] = events_copy["x"] + np.random.randint(-self.f_shift, self.f_shift)
+        
+        # Delete out of bound events
+        events_copy = np.delete(
+            events_copy,
+            np.where(
+                (events_copy["x"] < 0) | (events_copy["x"] >= self.sensor_size[0])))
+        return events_copy
+
+class Blend:
+    def __init__(self, p_blend, sensor_size):
+        self.p_blend = p_blend
+        self.n_blend = 7644
+        self.sensor_size = sensor_size
+    def __call__(self, dataset: list, classes: list) -> list:
+        
+        for i in range(self.n_blend):
+            idx = np.random.randint(0,len(dataset))
+            idx2 = np.random.randint(0,len(classes[dataset[idx][1]]))
+            assert dataset[idx][1] == dataset[classes[dataset[idx][1]][idx2]][1]
+            dataset.append((self.blend(dataset[idx][0], dataset[classes[dataset[idx][1]][idx2]][0]), dataset[idx][1]))
+            
+        return dataset
+
+    def blend(self, X1, X2):
+        new_x= []
+        new_t= []
+        X1 = copy.deepcopy(X1)
+        X2 = copy.deepcopy(X2)
```

Comment:
you probably don't need this deep copy as you already deep copy the whole dataset

Reply:
instead deleted the copy in the training loop